### ISSUE TITLE
sum of matrix always results in a 1x1 matrix

### ISSUE THREAD

**Issue by MDeiml** [Daniel-Diaz/matrix#22]
If you apply the `sum` function to a list of nxm matrices you get a 1x1 matrix as result e.g. 
`sum [M.fromLists [[1,1]], M.fromLists [[2,2]]] == ( 3 )`

**Comment by Daniel-Diaz** [comment]
Right, and I guess this is cause by the ``sum`` function actually evaluating your sum as:

```
M.fromLists [[1,1]] + M.fromLists [[2,2]] + fromInteger 0
```

Function ``fromInteger`` will create the 1x1 matrix ``( 0 )``, and this is the reason you are getting a 1x1 matrix as a result.

I guess an alternative ``sum`` function for matrices will help?

**Comment by MDeiml** [comment]
Actually `M.fromLists [[1,1]] + M.fromLists [[2,2]] + fromInteger 0` crashes with

    getElem: Trying to get the (1,2) element from a 1x1 matrix.

What really gets evaluated is `fromInteger 0 + fromLists [[1,1]]`
Maybe `elementwise` should be rewritten to accept matrices of different size?

**Comment by MDeiml** [comment]
This should work:

    elementwise f m m' = matrix (max (nrows m) (nrows m')) (max (ncols m) (ncols m')) $
      \k -> f (m ! k) (m' ! k)

**Comment by Daste** [comment]
Pheraps creating a function like:
`
  elementWiseMonoid :: (Monoid a,Monoid b) => ( a -> b  -> c ) -> Matrix a -> Matrix b -> Matrix c
  elementWiseMonoid f m m' = elementWiseUnsafe f (setSize rows columns m) (setSize rows columns m') 
     where 
         rows = max ( nrows m ) ( nrows m' )
         columns = max ( ncols m ) ( ncols m' )
`
and using it to instantiate  Num could help.
I did not test that.
